net: gptp: fix follow_up timestamp and announce byte order

chencaidy opened the pull request:
This PR include gPTP accuracy optimize and some byte order fix.

1. Change follow_up message preciseOriginTimestamp from software trigger to hardware egress capture
2. Fix announce message `cur_utc_offset` and `offset_scaled_log_var` fields byte order.

Using mimxrt1050-evk board and Intel i210 ethernet card direct cable connect, host using linuxptp stack,
and test PPS phase between 1588_EVENT2_OUT and SDP1 by oscilloscope.

Before:
mimxrt1050-evk as grandmaster, i210 as slave, PPS offset is 3.5ms
i210 as grandmaster, mimxrt1050-evk as slave, PPS offset is 20-30us

After:
mimxrt1050-evk as grandmaster, i210 as slave, PPS offset is 540-580ns
i210 as grandmaster, mimxrt1050-evk as slave, PPS offset is 1250-1380ns

chencaidy committed:
drivers: ethernet: mcux: improve receive timestamp accuracy

When ENET_ENHANCEDBUFFERDESCRIPTOR_MODE enabled,
MAC will automatic capture receive nanosecond from 1588TMR
and return back to ENET_ReadFrame. It is a highest accuracy
recv timestamp_ns, we do not need manually read from 1588TMR.

By this change, receive timestamp accuracy increase
from 20us to 200ns above.

Signed-off-by: Chen Caidy <chen@caidy.cc>

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_messages.c:
```diff
@@ -267,14 +271,19 @@ struct net_pkt *gptp_prepare_follow_up(int port, struct net_pkt *sync)
 	hdr->domain_number = 0U;
-	/* Store timestamp value in correction field. */
-	hdr->correction_field = gptp_timestamp_to_nsec(&sync->timestamp);
+	/* Gramd master clock should keep correction_field at zero */
```
Typo s/Gramd/Grand/

jukkar commented on subsys/net/l2/ethernet/gptp/gptp_md.c:
```diff
@@ -38,36 +38,13 @@ static void gptp_md_follow_up_prepare(struct net_pkt *pkt,
 	hdr = GPTP_HDR(pkt);
```
The `net: gptp: fix follow_up timestamp and announce byte order` contains three fixes. Could you separate them to three commits?
Please also enhance the commit messages so that they answer the question "Why this change is needed". The current changes look somewhat ok but it would be nice to have a bit more information (for example links to spec etc would be really helpful if possible).

chencaidy replied:
Heys [jukkar](https://github.com/jukkar),
Already separated them to three commits, and add some reference in commit.
The document can be found at https://1.ieee802.org/wp-content/uploads/2019/03/802-1AS-rev-d8-0.pdf

chencaidy committed:
net: gptp: fix follow_up message timestamp

According to IEEE802.1AS 11.4.4.2.1, we need fill
preciseOriginTimestamp as syncEventEgressTimestamp.

In this follow_up message, prec_orig_ts need to filled
from net_pkt_timestamp(sync) for best accuracy. state machine
is software trigger with insufficient precision.

After this change, a grand master endpoint sync accuracy
increase from 3.5ms to 580ns with mimxrt1050_evk board.

Signed-off-by: Chen Caidy <chen@caidy.cc>

chencaidy committed:
net: gptp: fix follow_up message correction_field

According to IEEE802.1AS Table 11-6 and 10.6.2.2.9,
802.1AS using peer-to-peer delay mechanism, two-step clock,
Grand master clock should keep this correction_field as zero.

Signed-off-by: Chen Caidy <chen@caidy.cc>

chencaidy committed:
net: gptp: fix announce message byte order

Linuxptp report UTC offset is 9472 seconds, is a byte order issue,
The right value is 37. Also fixed offset_scaled_log_var byte order.

Signed-off-by: Chen Caidy <chen@caidy.cc>

jukkar reviewed (approved):
Thanks, LGTM

dleach02 commented:
@chencaidy, thank you for the contribution! 

@hakehuang Can you test this PR to see how it does against #33747?

chencaidy commented:
> @chencaidy, thank you for the contribution!
> 
> @hakehuang Can you test this PR to see how it does against #33747?

Here is my test report, may be able to offer some help.

OS: Ubuntu 22.04 with 5.19 kernel
Linuxptp version: [aa60db2](https://github.com/richardcochran/linuxptp/commit/aa60db270be12e7144e7b0cef2f5dde4213b7057)
testptp: https://elixir.bootlin.com/linux/v5.19/source/tools/testing/selftests/ptp/testptp.c

1. Build linuxptp and testptp (gcc -o testptp testptp.c)
2. Modity linuxptp/configs/gPTP.conf, `neighborPropDelayThresh 50000`, because mcux driver issue make path delay too high
3. Build gptp sample: west build -p always -b mimxrt1050_evk samples/net/gptp
4. If using mimxrt1050_evk as grandmaster, set `CONFIG_NET_GPTP_GM_CAPABLE=y` in prj.conf. using i120 as grandmaster, set to n, then flash to board.
5. Start host linuxptp: sudo ./ptp4l -f configs/gPTP.cfg -i `enp1s0` -m
6. Enable pps out in SDP1: sudo ./testptp -d /dev/ptp0 -L 1,2 -p 1000000000
If SDP1 pps out missing, rerun step 6.

Here is my test platform:
![platform](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/3adf49e9-278a-4cac-9792-891c9cd6b229)

- rt1052 as grandmaster before

![linuxptp](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/74ab1647-362a-4aac-9c4a-5b5f389c203c)
linuxptp: path delay >18000ns, rms > 2000ns, sometimes have a jerk.

![oscilloscope](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/cf98569c-c24e-4417-b11a-2cb612f6e844)

- rt1052 as grandmaster after

![linuxptp](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/4f4392ad-a74e-4b19-a31a-337c472665eb)
linuxptp: delay < 900ns, rms < 50ns, no jerk.

![oscilloscope](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/c557ddaf-e407-41a2-a4f6-6b103012e856)

- host i210 as grandmaster before

![oscilloscope](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/f9c7050c-1717-44a8-8a81-dd12686432ce)

- host i210 as grandmaster after

![oscilloscope](https://github.com/zephyrproject-rtos/zephyr/assets/18113342/5c195236-6331-4be7-a730-3efc1fd8ac41)

hakehuang commented:
@chencaidy please take a looks at https://github.com/zephyrproject-rtos/zephyr/pull/60855 which should fix the jitter issue, and sync the timestamp in 24 hours
